### PR TITLE
Feature/e2e telemetry

### LINKED ISSUE #2330: Telemetry to report incidence of UISIs
https://github.com/vector-im/riot-meta/issues/187

### PR DESCRIPTION
**Pull request by ganfra**
Resolve #2330

### CODE REVIEW

**Comment by bmarty** [comment]
Warning, the merge will not compile, `liveState` has been removed from SDK

File: vector/src/main/java/im/vector/analytics/Events.kt [hunk 1]
```diff
@@ -24,10 +28,16 @@ enum class Action(val value: String) {
  * @param title the title associated with the event
  * @param value the optional value associated with the event
  */
-sealed class Event(val category: Category, val action: Action, val title: String? = null, val value: Float? = null) {
-    data class InitialSync(val duration: Long) : Event(Category.METRICS, Action.STARTUP, "initialSync", duration.toFloat())
-    data class IncrementalSync(val duration: Long) : Event(Category.METRICS, Action.STARTUP, "incrementalSync", duration.toFloat())
-    data class StorePreload(val duration: Long) : Event(Category.METRICS, Action.STARTUP, "storePreload", duration.toFloat())
-    data class LaunchScreen(val duration: Long) : Event(Category.METRICS, Action.STARTUP, "launchScreen", duration.toFloat())
-    data class Rooms(val nbOfRooms: Int) : Event(Category.METRICS, Action.STATS, "rooms", nbOfRooms.toFloat())
+sealed class TrackingEvent(val category: Category, val action: Action, val title: String? = null, val value: Float? = null) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/analytics/Events.kt]: it seems that default value are never used and parameters are never null. Simplify signature?

File: vector/src/main/java/im/vector/util/RiotEventDisplay.java [hunk 2]
```diff
@@ -100,6 +104,12 @@ public CharSequence getTextualDisplay(Integer displayNameColor) {
             } else {
                 text = super.getTextualDisplay(displayNameColor);
             }
+            if (mEvent.getCryptoError() != null) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/util/RiotEventDisplay.java]: Please take the opportunity to add missing @override annotation to this method

File: vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                }
+    }
+
+    //endregion
+
+    companion object {
+        const val CHECK_PERIOD = 5 * 1000L
+        // Give events a chance to be decrypted by waiting `GRACE_PERIOD` before counting
+        // and reporting them as failures (in ms)
+        const val GRACE_PERIOD = 60 * 1000L
```
Comment (bmarty) [inline vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt]: private consts?

File: vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    /**
+     * Reacts when an event is decrypted.
+     * Potentially removes a previously reported failure
+     * @param event the event who is decrypted
+     */
+    override fun onEventDecrypted(event: Event?) {
+        reportedFailures.remove(event?.eventId)
+    }
+
+    //region Private methods
```
Comment (bmarty) [inline vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt]: iOS Style :)

File: vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+     */
+    fun reportUnableToDecryptError(event: Event, roomState: RoomState, userId: String) {
+        if (reportedFailures[event.eventId] != null || trackedEvents.contains(event.eventId)) {
+            return
+        }
+        // Filter out "expected" UTDs
+        // We cannot decrypt messages sent before the user joined the room
+        val myUser = roomState.getMember(userId)
+        if (myUser.membership != RoomMember.MEMBERSHIP_JOIN) {
+            return
```
Comment (bmarty) [inline vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt]: Maybe we should report this type of error with a specific DecryptionFailureReason type? @manuroe ?

File: vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    private val checkFailuresTimer = timer(period = CHECK_PERIOD) {
+        checkFailures()
+    }
+
+    /**
+     * Reports the decryption error to the tracker.
+     * The error will be filtered.
+     * @param event: The error event
+     * @param roomState: The state of the room when the decryption error happened
+     * @param userId: The user id of the current session user
```
Comment (bmarty) [inline vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt]: remove last word?

File: vector/src/main/java/im/vector/analytics/e2e/DecryptionFailure.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
+/*
+ Copyright 2018 New Vector Ltd
```
Comment (bmarty) [inline vector/src/main/java/im/vector/analytics/e2e/DecryptionFailure.kt]: Copyright block is not like others

File: vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt [hunk 8]
```diff
@@ -0,0 +1,114 @@
+/*
+ Copyright 2018 New Vector Ltd
```
Comment (bmarty) [inline vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt]: Copyright block is not like others

File: vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                .eachCount()
+                .forEach { reason, count ->
+                    analytics.trackEvent(TrackingEvent.DecryptionFailure(reason, count))
+                }
+    }
+
+    //endregion
+
+    companion object {
+        const val CHECK_PERIOD = 5 * 1000L
```
Comment (bmarty) [inline vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt]: Please document CHECK_PERIOD

File: vector/src/main/java/im/vector/analytics/Events.kt (hunk 1 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/analytics/Events.kt]: The title and value are optional for majority of Analytics solutions, including Piwik. It seems legit to keep that in mind for Events.

File: vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt (hunk 4 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt]: It does exist for Android Studio :)

File: vector/src/main/java/im/vector/analytics/Events.kt (hunk 1 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/analytics/Events.kt]: ok

File: vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt (hunk 4 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt]: ok

File: vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt (hunk 5 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/analytics/e2e/DecryptionFailureTracker.kt]: N/A (discussed IRL)

**Pull request merged**